x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,177 @@
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // If there are remaining gamm shares, we must re-lock them.
-	// remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
-	// if !remainingGammShares.IsZero() {
-	// 	newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-	// 	newLockId = newLock.ID
-	// 	if err != nil {
-	// 		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 	}
-	// 	// If the lock was unlocking, we begin the unlock from where it left off.
-	// 	if preUnlockLock.IsUnlocking() {
-	// 		_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
-	// 		if err != nil {
-	// 			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 		}
-	// 	}
-	// }
-
-	return positionId, amount0, amount1, liquidity, joinTime, poolIdLeaving, poolIdEntering, newLockId, nil
+func (k Keeper) MigrateLockedPositionFromBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, poolIdLeaving, poolIdEntering, gammLockId, concentratedLockId uint64, err error) {
+	// Get the balancer poolId by parsing the gamm share denom.
+	poolIdLeaving = gammtypes.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
+
+	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
+	poolIdEntering, err = k.gk.GetLinkedConcentratedPoolID(ctx, poolIdLeaving)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Get the concentrated pool from the provided ID and type cast it to ConcentratedPoolExtension.
+	concentratedPool, err := k.clk.GetPoolFromPoolIdAndConvertToConcentrated(ctx, poolIdEntering)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Check that lockID corresponds to sender, and contains correct denomination of LP shares.
+	lock, err := k.validateLockForUnpool(ctx, sender, poolIdLeaving, lockId)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	gammSharesInLock := lock.Coins[0]
+	preUnlockLock := *lock
+
+	// Before we break the lock, we must note the time remaining on the lock.
+	remainingLockTime := k.getExistingLockRemainingDuration(ctx, lock)
+
+	// Check if the lock has a corresponding synthetic lock.
+	synthLockBeforeMigration := k.lk.GetAllSyntheticLockupsByLockup(ctx, lockId)
+
+	// If it does, check if it is superfluid delegated or undelegating.
+	wasSuperfluidUndelegatingBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superunbonding")
+	wasSuperfluidDelegatedBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superbonding")
```

Comment:
Can we have a check that you can't be both at the same time?

Also, is there ever a case where synth lock can be neither?

Reply:
These are the only two possibilities, as shown here

https://github.com/osmosis-labs/osmosis/blob/ed085524d91cc472ec20eb3795dcae833c3c5de1/x/superfluid/keeper/synthetic_lock_wrapper.go#L12-L17

I can add a defense in depth measure though to be sure